base_local_planner/src/obstacle_cost_function.cpp

```diff
@@ -62,23 +62,25 @@ ObstacleCostFunction::~ObstacleCostFunction() {
   }
 }
 
-
-void ObstacleCostFunction::setParams(double max_trans_vel, double max_forward_inflation, double max_sideward_inflation, double scaling_speed, bool occdist_use_footprint) {
+void ObstacleCostFunction::setParams(double max_trans_vel, double max_forward_inflation, double max_sideward_inflation,
+                                     double scaling_speed, bool occdist_use_footprint, double scaling_discount_factor) {
   // TODO: move this to prepare if possible
   max_trans_vel_ = max_trans_vel;
   max_forward_inflation_ = max_forward_inflation;
   max_sideward_inflation_ = max_sideward_inflation;
   scaling_speed_ = scaling_speed;
   occdist_use_footprint_ = occdist_use_footprint;
+  scaling_discount_factor_ = scaling_discount_factor;
 }
 
 void ObstacleCostFunction::setFootprint(std::vector<geometry_msgs::Point> footprint_spec) {
   footprint_spec_ = footprint_spec;
 }
 
-std::vector<geometry_msgs::Point> ObstacleCostFunction::getScaledFootprint(const Trajectory& traj) const {
+std::vector<geometry_msgs::Point> ObstacleCostFunction::getScaledFootprint(const Trajectory& traj, unsigned int index) const {
+  index = std::min(index, traj.getPointsSize() - 1);
   std::vector<geometry_msgs::Point> scaled_footprint = footprint_spec_;
-  const double scale = getScalingFactor(traj, scaling_speed_, max_trans_vel_);
+  const double scale = std::pow(scaling_discount_factor_, index) * getScalingFactor(traj, scaling_speed_, max_trans_vel_);
```

Comment:
S: make a lookup table since the scaling will not change. Avoiding computing all `pow` on each loop for each trajectory. e.g. `scaling_discount_factor_vector[index]` (compute this in reconfigure)

Reply:
I think Renan has a point here; and if you use a vector instead of unordered_map, it's around 10 times faster to pre-compute: https://godbolt.org/z/7zrqv7xe5

Reply:
ok yeah using vector makes more sense here :facepalm: 

But do we really care about saving ~0.4ms off when our code runs at 15hz (i.e. 66ms each loop)?
I don't really think the added complexity is worth to save <1% execution time...
If you both want it I'll implement the change though :shrug:

Reply:
yes, I was thinking using vectors...
And I don't think it is uglier, but I am ok doing it later if we have a task for profiling.

Reply:
Btw on Friday when profiling I compared this PR with devel and the result was the same